### PR TITLE
[mlir] [mem2reg] Fix Mem2Reg attempting to promote in graph regions

### PR DESCRIPTION
**Pull request by Moxinilian**
Mem2Reg assumes SSA dependencies but did not check for graph regions. This fixes it.

### CODE REVIEW

File: mlir/lib/Transforms/Mem2Reg.cpp [hunk 1]
```diff
@@ -255,6 +256,18 @@ LogicalResult MemorySlotPromotionAnalyzer::computeBlockingUses(
   // delete itself). We thus need to start from the use of the slot pointer and
   // propagate further requests through the forward slice.
 
+  // Because this pass currently only supports analysing the parent region of
+  // the slot pointer, if a promotable memory op that needs promotion is within
+  // a graph region, the slot may only be used in a graph region and should
+  // therefore be ignored.
+  Region *slotPtrRegion = slot.ptr.getParentRegion();
+  auto slotPtrRegionOp =
+      llvm::dyn_cast<RegionKindInterface>(slotPtrRegion->getParentOp());
```
Comment (Dinistro) [inline mlir/lib/Transforms/Mem2Reg.cpp:265]:
```suggestion
     dyn_cast<RegionKindInterface>(slotPtrRegion->getParentOp());
```

File: mlir/test/Transforms/mem2reg.mlir [hunk 2]
```diff
@@ -1,4 +1,4 @@
-// RUN: mlir-opt %s --pass-pipeline='builtin.module(func.func(mem2reg))' --split-input-file | FileCheck %s
+// RUN: mlir-opt %s --pass-pipeline='builtin.module(func.func(mem2reg),test.isolated_graph_region(mem2reg))' --split-input-file | FileCheck %s
```
Comment (Dinistro) [inline mlir/test/Transforms/mem2reg.mlir:1]:
```suggestion
// RUN: mlir-opt %s --pass-pipeline='builtin.module(any(mem2reg))' --split-input-file | FileCheck %s
```
This this not work?

**Review by Dinistro — APPROVED** [review]
LGTM % a nits. Thanks for the improvement.

File: mlir/test/Transforms/mem2reg.mlir (hunk 2 above)
Comment (Moxinilian) [reply mlir/test/Transforms/mem2reg.mlir]: Just checked, it works, thanks!

**Pull request merged**